File: mark.py
#!/usr/bin/env python
#
# Copyright (C) 2010-2012 John Driscoll <johnoliverdriscoll@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# QTPy Mark class

import move

class Mark:
  """ Represents a player's mark on the game board
  
  Attributes:
    type: Spooky or Classic
    square: Square this mark resides in
    link: If this is a spooky mark, this is the other mark on the board
    player: Player number 1 or 2
    weight: The mark's weight (turn number)
  """
  
  def __init__(self,
               square,
               type=move.CLASSIC,
               player=1,
               weight=1,
               link=None):
    self.square = square
    self.type = type
    self.player = player
    self.weight = weight
    self.link = link
  
  def go_classic(self):
    """ Converts a spooky mark into a classic piece and causes entangled
    marks to collapse """
    
    marks = []
    self.square._spookies = []
    for mark in self.square.marks: # back up other marks in this square
      self.square._spookies.append(mark)
      if mark != self: marks.append(mark)
    # Put a classic mark in the square
    self.square.marks = [ Mark(self.square, type=move.CLASSIC,
                               player=self.player, weight=self.weight) ]
    # Force entangled marks to turn their links into classic moves
    for mark in marks:
      # Prevent reading over a cycle again
      if not mark.link.square.is_classic():
        mark.link.go_classic()
  
  def go_spooky(self):
    """ Replace classic piece with spooky and uncollapses squares to the
    entangled state """
    
    self.square.marks = self.square._spookies
    for mark in self.square.marks:
      if mark.link.square.is_classic():
        mark.link.go_spooky()

File: static/js/mark.js
/*
 * QTPy Mark class
 *
 * John Driscoll
 */

function Mark() {
  
  this.goClassic = function() {
    // Converts a spooky mark into a classic piece and causes entangled marks to collapse
    
    var marks = [],
        self = this;
    // back up other marks in this square
    for (var i = 0; i < this.square.marks.length; i++) {
      var mark = this.square.marks[i];
      if (mark != this) marks.push(mark);
    }
    // Put a classic mark in the square
    this.square.marks = [ new Mark(this.square, CLASSIC, this.player, this.weight) ];
    // Force entangled marks to turn their links into classic moves
    for (var i = 0; i < marks.length; i++) {
      var mark = marks[i];
      if (!mark.link.square.isClassic()) // Prevent reading over a cycle again
        mark.link.goClassic();
    }
    if (!this.classic) this.animateCollapse();
  };
  
  this.destroy = function() {
    // Destroy marks that will never be played
    
    if (this.spooky) {
      this.spooky.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
      this.orbit.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
      this.line.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
    }
  };
  
  this.animateHide = function() {
    if (this.spooky) {
      this.spooky.stop();
      this.spooky.orbit.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
      this.spooky.link.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
    }
    if (this.link && this.link.spooky) {
      this.link.spooky.orbit.animate({ opacity: 0 }, 500, null, function() { this.hide(); });
      this.link.spooky.animate({ r: 0 }, 500, null, function() { this.hide(); });
    }
  };
  
  this.animateCollapse = function() {
    // Animate
    var fill = qtpy.color(this.player),
        stroke = fill,
        cx = this.square.rect.attr('x') + this.square.rect.attr('width') / 2,
        cy = this.square.rect.attr('y') + this.square.rect.attr('height') / 2,
        m = ra.circle(this.spooky ? this.spooky.attr('cx') : cx,
                      this.spooky ? this.spooky.attr('cy') : cy,
                      this.spooky ? this.spooky.attr('r') : 3).attr({
              fill: fill,
              stroke: stroke
            }).behindBoard();
    this.classic = true;
    // Don't animate squares that have 1 possible classical mark, because the classical mark is
    // already there
    this.animateHide();
    if (this.square.marked) {
      if (this.spooky) this.spooky.hide();
      return;
    }
    m.animate({
      cx: cx,
      cy: cy,
      r: 35
    }, 500, 'bounce', function() { this.stop() });
    // Create new circle for player mark (prevent animation conflict that causes orbiting classical)
    this.animation = m;
    var self = this;
    safeTimeout($game, function() {
      if (self.spooky)
        self.spooky.hide();
      self.animation.hide();
      self.animation.stop();
      ra.circle(cx, cy, 35).attr({
        fill: fill,
        stroke: stroke
      });
    }, 500);
    // Hide original spooky
    if (!this.square.marked && this.spooky)
      this.spooky.hide();
    else if (this.spooky) {
      // marked square
      this.spooky.link.animate({ opacity: 0 }, 500);
    }
    if (this.square.spooky) {
      // Don't even know how a spooky became attached to a square directly.
      this.square.spooky.stop().hide();
      this.square.spooky.orbit.hide();
    }
  };
  
  // init
    this.square = arguments[0];
    this.type = arguments[1] || CLASSIC;
    this.player = arguments[2] || 1;
    this.weight = arguments[3] || 1;
    this.link = arguments[4] || null;
}